### ISSUE TITLE
Ubuntu build

### ISSUE THREAD

**Issue by ASnow** [jbreeden/rubium#3]
Try build for linux. `rake cef:cefclient` command failed with `/usr/bin/ld: cannot find -lXXX`.
Can you help me?

**Comment by jbreeden** [comment]
Sure thing! Sorry, I don't have good build instructions or scripts yet. I'm sure we can figure it out though.

Are you following the notes from README/rubium_build.linux.txt?

If so, the comment before `rake cef:cefclient` is really an instruction to the user, not a comment on what's about to happen (I realize this isn't super clear). Did you already download & extract CEF as described there? (EDIT: Added some extra comments in that file for now. May be helpful... may not.)

I'll run through the build from scratch on a fresh Ubuntu install after work to see if any of the notes are out of date.

**Comment by ASnow** [comment]
I figure out that preveously try to build x86_64 CEF on linux i386, it was stupid.
Now I stuck with mruby

```
$ rake mruby:build --trace
** Invoke mruby:build (first_time)
** Execute mruby:build
rake
GIT CHECKOUT master 
Already on 'master'
Your branch is up-to-date with 'origin/master'.
GIT CHECKOUT master 
Already on 'master'
Your branch is up-to-date with 'origin/master'.
GIT CHECKOUT master 
Already on 'master'
Your branch is up-to-date with 'origin/master'.
rake aborted!
Don't know how to build task '/home/asnow/dev/apr/rubium/mruby/build/host/mrbgems/mruby-apr/src/mruby_apr_vformatter_buff_t.o'

Tasks: TOP => default => all => /home/asnow/dev/apr/rubium/mruby/build/host/lib/libmruby.flags.mak => /home/asnow/dev/apr/rubium/mruby/build/host/lib/libmruby.a
(See full trace by running task with --trace)
rake aborted!
Command failed with status (1): [rake...]
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/file_utils.rb:53:in `block in create_shell_runner'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/file_utils.rb:45:in `call'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/file_utils.rb:45:in `sh'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/file_utils_ext.rb:39:in `sh'
/home/asnow/dev/apr/rubium/Rakefile.rb:24:in `block (3 levels) in <top (required)>'
/home/asnow/dev/apr/rubium/Rakefile.rb:22:in `chdir'
/home/asnow/dev/apr/rubium/Rakefile.rb:22:in `block (2 levels) in <top (required)>'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/task.rb:205:in `call'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/task.rb:205:in `block in execute'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/task.rb:200:in `each'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/task.rb:200:in `execute'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/task.rb:158:in `block in invoke_with_call_chain'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/monitor.rb:211:in `mon_synchronize'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/task.rb:151:in `invoke_with_call_chain'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/task.rb:144:in `invoke'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:116:in `invoke_task'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:94:in `block (2 levels) in top_level'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:94:in `each'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:94:in `block in top_level'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:133:in `standard_exception_handling'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:88:in `top_level'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:66:in `block in run'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:133:in `standard_exception_handling'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/lib/ruby/1.9.1/rake/application.rb:63:in `run'
/home/asnow/.rvm/rubies/ruby-1.9.3-p551/bin/rake:32:in `<main>'
Tasks: TOP => mruby:build​
```

**Comment by jbreeden** [comment]
That's strange. Seems like it's finding the .c file and is able to create the .o file task... then loses track of it somehow...

First thing I would try is 

```
rake mruby:clean && rake mruby:build
```

Assuming you're past that already, I would try to eliminate some other variables. That file is in the mruby-apr gem, so I would isolate that gem and try building mruby & it alone.

You can do that by editing the mruby/build_config.rb file and commenting out the lines shown below (just leave mruby-ap, the call to configure_mruby_apr is only needed when the other gems are used)

```
#  conf.gem :github => 'iij/mruby-regexp-pcre'
  conf.gem :github => "jbreeden/mruby-apr"
#  conf.gem :github => "jbreeden/mruby-sqlite"
#  configure_mruby_apr(conf)
#  conf.gem "../mrbgems/mruby-rubium"
#  conf.gem "../mrbgems/mruby-bin-rubium"
#  conf.gem "../mrbgems/mruby-cef"
```

Once that is done, build using mruby's rake tasks, instead of the one that comes with rubium (it sets some other compiler flags that may be throwing your build off). 

```
cd path/to/rubium/mruby
rake clean
rake
```

You could also update mruby... just keep the same build_config.rb

```
cd path/to/rubium/mruby
git remote add upstream https://github.com/mruby/mruby
git pull upstream master
git checkout HEAD build_config.rb
rake clean
rake
```

The build system should already be pulling down the latest mruby-apr, so nothing to do there.

Hopefully that narrows it down. Let me know what you find!

**Comment by jbreeden** [comment]
Finally got a chance to try the Linux build again. Looks like some recent changes in mruby-apr are causing problems (though, not the same one you're seeing). Working through those now.

**Comment by ASnow** [comment]
I think apr is configured for x64 platform

```
CC    build/mrbgems/mruby-apr/src/mruby_apr_vformatter_buff_t.c -> build/host/mrbgems/mruby-apr/src/mruby_apr_vformatter_buff_t.o
In file included from /home/asnow/dev/apr/rubium/mruby/build/mrbgems/mruby-apr/src/mruby_apr_vformatter_buff_t.c:10:
In file included from /home/asnow/dev/apr/rubium/mruby/build/mrbgems/mruby-apr/include/mruby_APR.h:801:
/usr/local/apr/include/apr-1/apr.h:358:10: error: unknown type name 'off64_t'
typedef  off64_t           apr_off_t;
         ^
/usr/local/apr/include/apr-1/apr.h:613:2: error: no decision has been made on APR_PATH_MAX for your platform
#error no decision has been made on APR_PATH_MAX for your platform
 ^
2 errors generated.
In file included from /home/asnow/dev/apr/rubium/mruby/build/mrbgems/mruby-apr/src/mruby_apr_vformatter_buff_t.c:10:
In file included from /home/asnow/dev/apr/rubium/mruby/build/mrbgems/mruby-apr/include/mruby_APR.h:801:
/usr/local/apr/include/apr-1/apr.h:358:10: error: unknown type name 'off64_t'
typedef  off64_t           apr_off_t;
         ^
/usr/local/apr/include/apr-1/apr.h:613:2: error: no decision has been made on APR_PATH_MAX for your platform
#error no decision has been made on APR_PATH_MAX for your platform
 ^
2 errors generated.
rake aborted!
Command failed with status (1): [clang -g -std=gnu99 -O3 -Wall -Werror-impl...]
/home/asnow/dev/apr/rubium/mruby/tasks/mruby_build_commands.rake:31:in `_run'
/home/asnow/dev/apr/rubium/mruby/tasks/mruby_build_commands.rake:36:in `rescue in _run'
/home/asnow/dev/apr/rubium/mruby/tasks/mruby_build_commands.rake:32:in `_run'
/home/asnow/dev/apr/rubium/mruby/tasks/mruby_build_commands.rake:88:in `run'
/home/asnow/dev/apr/rubium/mruby/tasks/mruby_build_commands.rake:112:in `block (2 levels) in define_rules'
/home/asnow/.rvm/gems/ruby-2.2.1@global/bin/ruby_executable_hooks:15:in `eval'
/home/asnow/.rvm/gems/ruby-2.2.1@global/bin/ruby_executable_hooks:15:in `<main>'
Command failed with status (1): ["clang" -g -std=gnu99 -O3 -Wall -Werror-im...]
/home/asnow/dev/apr/rubium/mruby/tasks/mruby_build_commands.rake:33:in `_run'
/home/asnow/dev/apr/rubium/mruby/tasks/mruby_build_commands.rake:88:in `run'
/home/asnow/dev/apr/rubium/mruby/tasks/mruby_build_commands.rake:112:in `block (2 levels) in define_rules'
/home/asnow/.rvm/gems/ruby-2.2.1@global/bin/ruby_executable_hooks:15:in `eval'
/home/asnow/.rvm/gems/ruby-2.2.1@global/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => default => all => /home/asnow/dev/apr/rubium/mruby/build/host/lib/libmruby.flags.mak => /home/asnow/dev/apr/rubium/mruby/build/host/lib/libmruby.a => /home/asnow/dev/apr/rubium/mruby/build/host/mrbgems/mruby-apr/src/mruby_apr_vformatter_buff_t.o
(See full trace by running task with --trace)
```

but I can't understand how to build for 32bits.

**Comment by jbreeden** [comment]
Are you building with rubium or just mruby & mruby-apr? I know I forced 64bit for some platforms with rubium due to Chromium Embedded restrictions.

**Comment by jbreeden** [comment]
Good news...

I have a fix for the "#error no decision has been made on APR_PATH_MAX for your platform". I'll merge into master shortly (update: just merged [90066e0](https://github.com/jbreeden/mruby-apr/commit/90066e01ef74681b0066403139c3e8216f20f1e0)), but I think you can just remove the "-std=c11" lines from mruby-apr's mrbgem.rake

As for the off64_t, see if this helps: http://stackoverflow.com/questions/22663897/unknown-type-name-off64-t

**Comment by jbreeden** [comment]
Standing up a 32bit VM now, so we should be on the same page soon.

**Comment by jbreeden** [comment]
Able to build mruby w/ mruby-apr on 32bit linux with a checkout of the current head:

```
rake clean default CFLAGS="-D_LARGEFILE64_SOURCE"
```

All the unit tests passed, so that's a good sign.

The `-D_FILE_OFFSET_BITS=64` flag is documented as the preferred method for new programs, but had no effect when I tried on the command line. Maybe I did something else wrong...

I'll try to finish building rubium and see what else happens.

**Comment by jbreeden** [comment]
Success!

I've pushed [230678e](https://github.com/jbreeden/rubium/commit/230678e24a10014d9258d8c2b32172de413aafff) to rubium. That should make sure you can build with whatever CEF version you drop in (no longer tied to 64-bit or a specific version).

I also updated the mruby submodule to pull the latest from upstream mruby.

As long as you use the -D_FILE_OFFSET_BITS=64 flag when you rake mruby...

```
rake mruby:build CFLAGS="-D_LARGEFILE64_SOURCE"
```

You should be good to go!

Just a note, when you run `rake package` you should get a folder called "rubium" in your "rubium" directory, that contains the "rubium" executable (I could have chose my names better). Export this folder on your path, then try:

```
cd /path/to/rubium/samples/mruby-edit
rubium --dev-tools
```

You should get a chrome window with an Ace code editor & command line. There should be another window with the chrome dev tools.

**Comment by ASnow** [comment]
@jbreeden, it's work. Great! Thanks!
